### ISSUE TITLE
Feature Request: window-copy without losing scrollback position

### ISSUE THREAD

**Issue by atweiden** [tmux/tmux#402]
Credit @alps2006 for this patch which prevents tmux from losing your place in the scrollback after copying text: https://github.com/tmux-plugins/tmux-yank/issues/40#issuecomment-218341560

Instead of forcing your cursor down to the bottom of the scrollback after copying, your cursor stays in the same position (can always hit <kbd>Enter</kbd> to scroll to the bottom if that's what you want).

```diff
diff --git a/window-copy.c b/window-copy.c
index b352d5a..af36be3 100644
--- a/window-copy.c
+++ b/window-copy.c
@@ -690,8 +690,8 @@ window_copy_key(struct window_pane *wp, struct client *c, struct session *sess,
 	case MODEKEYCOPY_COPYPIPE:
 		if (sess != NULL) {
 			window_copy_copy_pipe(wp, sess, NULL, arg);
-			window_pane_reset_mode(wp);
-			return;
+			window_copy_clear_selection(wp);
+			window_copy_redraw_screen(wp);
 		}
 		break;
 	case MODEKEYCOPY_COPYSELECTION:
```

It seems to work, and is a big improvement over the default of forcing your cursor down to the bottom of the scrollback.

I'm perfectly happy to maintain this feature as a patch, but I thought others might be interested in having this be the default behavior in tmux.

I should add it would be a further marked improvement if the same behavior would be corrected when doing mouse selections.

**Comment by nicm** [comment]
It is not an improvement and in any case is already supported by using -x.

Credit @alps2006 <https://github.com/alps2006> for this patch which
prevents tmux from losing your place in the scrollback after copying
text: tmux-plugins/tmux-yank#40
(comment)
<https://github.com/tmux-plugins/tmux-yank/issues/40#issuecomment-218341560>

Instead of forcing your cursor down to the bottom of the scrollback after
copying, your cursor stays in the same position (can always hit Enter to
scroll to the bottom if that's what you want).

diff --git a/window-copy.c b/window-copy.c
index b352d5a..af36be3 100644--- a/window-copy.c+++ b/window-copy.c@@
-690,8 +690,8 @@ window_copy_key(struct window_pane *wp, struct client
*c, struct session *sess,
    case MODEKEYCOPY_COPYPIPE:
        if (sess != NULL) {
            window_copy_copy_pipe(wp, sess, NULL, arg);-
window_pane_reset_mode(wp);-           return;+
window_copy_clear_selection(wp);+
window_copy_redraw_screen(wp);
        }
        break;
    case MODEKEYCOPY_COPYSELECTION:

It seems to work, and is a big improvement over the default of forcing your
cursor down to the bottom of the scrollback.

I'm perfectly happy to maintain this feature as a patch, but I thought
others might be interested in having this be the default behavior in tmux.

I should add it would be a further marked improvement if the same behavior
would be corrected when doing mouse selections.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub
<https://github.com/tmux/tmux/issues/402>

**Comment by atweiden** [comment]
Now that I see the alternative was this easy:

```tmux.conf
bind-key -t vi-copy y copy-pipe -x "xsel -i -p -b"
```

I'll close this.

**Comment by atweiden** [comment]
Whoops, I had forgotten to recompile before closing the issue. It appears that `copy-pipe -x` is not implemented. Can you explain how it's possible to achieve the desired effect without the proposed patch?

**Comment by nicm** [comment]
It is not supported for copy-pipe.

**Comment by atweiden** [comment]
Without this patch then, how is it possible to copy a visual selection to clipboard without losing your place in the scrollback?

**Comment by nicm** [comment]
You can use copy-selection -x; you can't use -x with copy-pipe.

It would be good if that worked too, but mode key bindings currently
only support one argument and for copy-pipe that is needed for the pipe
command, so we would need support for two or more arguments.

This patch is not enough, because we don't want to change the existing
behaviour for everyone. Adding -x to copy-pipe as well is already in the
TODO file.

**Comment by timotheecour** [comment]
This is relevant:

https://github.com/tmux/tmux/issues/337#issuecomment-198872468
[in mouse mode, after selecting text in a scrolled window, it scrolls to
end]

nicm@ proposed using:

bind -temacs-copy MouseDragEnd1Pane copy-selection -x

however this isn't great: see reason in that post, quoted below:

actually the proposed fix is not great and I now see why you don't want it
to be the default.
what i would like (as many others, see +1's under #140
<https://github.com/tmux/tmux/issues/140>):

when scrolling, upon releasing a text selection, it stays in scrolled mode
(and hence copy mode)

when not scrolling, upon releasing a text selection, it goes back to normal
mode (ie the current default).

in other words, selecting text doesn't change the state (the fact that
selecting text enters copy mode is an implementation detail: it should exit
copy mode after selection only if we weren't in copy mode before mouse
selection)

this is IMO follows the principle of least surprise, and is exactly what
the others wanted too.




On Wed, May 11, 2016 at 1:26 PM, Nicholas Marriott <notifications@github.com

**Comment by nicm** [comment]
No. After you copy, you want to paste, so it should exit copy mode to
let you do so. Just sitting there is silly. After you press M-w in
emacs, does it sit there with text still selected? No. This is POLS.

**Comment by nicm** [comment]
These problems would all go away if copy mode just used a normal key table, like I intend to do... sometime...

**Comment by atweiden** [comment]
Well, the patch [is here](https://github.com/atweiden/pkgbuilds/commit/c1923f9d66a646c4aef3ea0c1df471a37f6b2e07) if anyone wants it. I'm glad to see this is on the list of todos.

**Comment by chikamichi** [comment]
@nicm hi :) what about:

> when not scrolling, upon releasing a text selection, it goes back to normal
mode (ie the current default).

I'm afraid I didn't understand what your take on this is exactly?

**Comment by mckellyln** [comment]
I use this (with master):
```
bind-key -T copy-mode-vi C-c send-keys -X copy-pipe "xclip -selection clipboard ; tmux send-keys -X clear-selection"
unbind -T copy-mode-vi MouseDragEnd1Pane
```
Seems to work ok.
(recent tmux source would require a different patch as a lot has changed)
There are workflows where you copy and then go to another pane or application and paste and then go back - so exiting copy-mode does not always make sense, at least for me :-)
Thank you for tmux, it is awesome!
